### ISSUE TITLE
Защита от ошибок

### ISSUE THREAD

**Issue by pansa** [tmp1test/t2#2]
"Предусмотреть обработку ошибок, которые могут возникнуть в процессе выполнения таких действий. - Готово"

Давайте рассмотрим, защиту от каких ошибок вы предусмотрели?

**Comment by pansa** [comment]
Чем больше защит от ошибок, тем лучше! =)
Давайте я дам наводящие вопросы, а вы сформулируете варианты, т.к, безусловно, мы ищем разработчика, который самостоятельно старается предусмотреть все подводные камни.

1. Какие нестандартные ситуации возможны со стороны клиента? В основном  это касается запросов на upload.

2. Какие могут быть сложности в работе реализованного сервиса, если хранилище на сервере будет содержать, допустим, 10 тб данных, и 2-3 десятка млн файлов?

3. Подумайте про то, что оперативная память сервера ограничена. Аккуратно ли расходует память ваш сервис?

**Comment by tmp1test** [comment]
@pansa 
1. Отправка файлов неавторизованным пользователем - добавлена проверка
Отправка больших файлов - есть проверка, которая разрывает соединение если Content-Length превышает максимальное значение размера файла
Также уже сделаны лимиты на одновременное количество соединений с ip, общее количество соединений и количество байт, загружаемых в сутки с одного ip

Возможно еще имеете в виду обрыв соединения? Загрузка больших файлов чанками?

2. Правильно ли я понимаю, что речь идет о масштабировании сервиса?
При таких объемах данных и должной нагрузке на одном сервере будут проблемы производительности. Тут может быть, что угодно - как cpu или память, так и входящий/исходящий канал, например.

Нужно думать, как шардировать и реплицировать данные между серверами. Анализировать метрики - каких запросов больше от юзеров? Вероятно, что основные проблемы будут на отдачу большого количества файлов, и понадобятся "слейвы" как раз для исходящего траффика.
Если важна быстрая доставка файлов из любой точки, то нужно смотреть в сторону CDN.
Также придется масштабировать и логику.

3. На данный момент нет. 
Например, при чтении файла он сначала читает файл в буфер, считает контрольную сумму по файлу и потом отдает его только клиенту.

Так сделано для упрощения кода. Можно не считать контрольную сумму каждый раз, хранить ее в метаданных и пересчитывать ее только, например, если время обновления файла отличается от даты создания. Либо какое-то другое решение.

При загрузке файла - аналогичная ситуация при проверке хешей. Тут тоже можно подумать над оптимизацией.

Других проблем не заметил )

**Comment by tmp1test** [comment]
@pansa 
извиняюсь, не все отписал

1. возможно имеется также ситуация одновременной загрузки одних и тех же файлов на сервер.
можно выдавать ошибку, если такой файл уже кем-то загружается в данный момент.

2. Под масштабированием логики понимаю:
- использование кластера редиса
- подключение распределенного рейт лимитера
- логика шардирования файлов между серверами
- автоочистка на нескольких серверах

**Comment by pansa** [comment]
> Возможно еще имеете в виду обрыв соединения? Загрузка больших файлов чанками?

Да, конечно, в т.ч. это.  Большие файлы не редкость, а сервис считывает их целиком в память.
А при обрыве соединения в хранилище останется "огрызок" файла. 

>Тут может быть, что угодно - как cpu или память, так и входящий/исходящий канал, например

С каналом всё понятно, это очевидно. 
Но разве объем данных должен влиять на потребляемую сервисом память? А CPU? 
Это еще может зависеть от кол-ва запросов в единицу времени, но кол-во данных на это не влияют. У вас, например, вычисление суммарного размера хранилища выполняется stat'ми по всем файлам. Когда файлов будет много, сервис просто "залипнет" на io, причем, ни cpu ни память он потреблять практически не будет.

> возможно имеется также ситуация одновременной загрузки одних и тех же файлов на сервер.
можно выдавать ошибку, если такой файл уже кем-то загружается в данный момент.

Не очень удачное решение. Почему ошибка? Предложите другой вариант?

> использование кластера редиса

Хм. А привидите кейс, когда баттлнеком может быть редис?

**Comment by tmp1test** [comment]
@pansa 

> У вас, например, вычисление суммарного размера хранилища выполняется stat'ми по всем файлам. Когда файлов будет много, сервис просто "залипнет" на io, причем, ни cpu ни память он потреблять практически не будет

Можно распараллелить процесс подсчета как минимум по папкам.
Также могу посчитать по мета-данным редиса. Но там не нравится запрос на получение всех ключей по паттерну. Видимо нужно изменить структуру хранения данных.
Какой вариант по вашему предпочтительнее? 

> Не очень удачное решение. Почему ошибка? Предложите другой вариант?

Тут в любом случае возможна гонка. При сохранении файла нужно ставить блокировку, второй клиент должен ждать некий таймаут, в течении которого будет несколько раз проверять - не загрузился ли уже файл другим клиентом. Если да, то все хорошо и мы просто отдаем его хеш.

Если нет, то то есть вариант грузить сразу оба файла в предхранилище. По окончании загрузки перемещать его в основное, тогда можно грузить сразу хоть 10 одинаковых файлов.

Подойдет такой вариант?

> Хм. А привидите кейс, когда баттлнеком может быть редис?

В данном случае наверное никогда. Но вообще редис юзает только один CPU. При большом rps и больших записях в него. Редис может показывать в своем слоулог, что у него все хорошо и он быстро отдает ответы, а на деле у него будет большая очередь и запросы отваливаются по таймаутам. Через netstat можно это увидеть.

**Comment by pansa** [comment]
> Можно распараллелить процесс подсчета как минимум по папкам.

... И убьете дисковую систему, которая и так у нас не справляется с ответами на stat. Увы, поиск по развеститым файловым системам - крайне медленный.

> Какой вариант по вашему предпочтительнее?

Ну мы же сейчас не разрабатываем реальный сервис. Нам интересны ваши варианты решения задачи =)

> клиент должен ждать некий таймаут, в течении которого будет несколько раз проверять - не загрузился ли уже файл другим клиентом. Если да, то все хорошо и мы просто отдаем его хеш.

Ну такой вариант, конечно, имеет право на жизнь. Но мы вынуждаем клиента чего-то ждать. 

> Если нет, то то есть вариант грузить сразу оба файла в предхранилище. По окончании загрузки перемещать его в основное

Именно так!

> Но вообще редис юзает только один CPU. 
Это да, но мы сейчас обсуждаем конкретный сервис.  Вы предложили кластеризовать редис. Я хотел узнать - какую проблему вы предполагаете этим решать? В какой ситуации нам пришлось бы это делать?

**Comment by tmp1test** [comment]
> Ну мы же сейчас не разрабатываем реальный сервис. Нам интересны ваши варианты решения задачи =)

Добавил общий счетчик количества загруженных байтов в редис.
При загрузке файла он инкрементируется, при удалении - декрементируется. Добавил тесты.

В таком виде все должно работать быстро, возможны расхождения в случае удаления файла не через API. Нужно периодически синхронизировать счетчик.

**Comment by tmp1test** [comment]
@pansa 

> Это да, но мы сейчас обсуждаем конкретный сервис. Вы предложили кластеризовать редис. Я хотел узнать - какую проблему вы предполагаете этим решать? В какой ситуации нам пришлось бы это делать?

Если бы мы делали распределенный рейт-лимитер на основе редиса (хранили бы счетчики с ttl). При достаточно большом rps, учитывая, что мы еще и пишем в этот же редис мета-данные файлов, мы могли бы как раз-таки упереться в проблему с очередью.
